src/uct/tcp/tcp_sockcm.c

```diff
@@ -36,18 +38,58 @@ static uct_cm_ops_t uct_tcp_sockcm_ops = {
     .listener_create  = UCS_CLASS_NEW_FUNC_NAME(uct_tcp_listener_t),
     .listener_reject  = uct_tcp_listener_reject,
     .listener_query   = uct_tcp_listener_query,
-    .listener_destroy = UCS_CLASS_DELETE_FUNC_NAME(uct_tcp_listener_t)
+    .listener_destroy = UCS_CLASS_DELETE_FUNC_NAME(uct_tcp_listener_t),
+    .ep_create        = UCS_CLASS_NEW_FUNC_NAME(uct_tcp_sockcm_ep_t)
 };
 
+static void uct_tcp_close_ep(uct_tcp_sockcm_ep_t *ep)
+{
+    ucs_list_del(&ep->list);
+    ucs_async_remove_handler(ep->fd, 1);
+    close(ep->fd);
+    ep->fd = -1;
+    UCS_CLASS_DELETE(uct_tcp_sockcm_ep_t, ep);
+}
+
 void uct_tcp_sa_data_handler(int fd, void *arg)
 {
-    ucs_warn("Function not implemented");
+    uct_tcp_sockcm_ep_t *ep = (uct_tcp_sockcm_ep_t*)arg;
+
+    ucs_assertv(ep->fd == fd, "ep->fd %d fd %d, ep_state %d", ep->fd, fd, ep->state);
+
+    if (ep->state & UCT_TCP_SOCKCM_EP_ON_SERVER) {
+        /* server side - the client disconnected and the server is waiting
+         * for EVREAD events */
+        uct_tcp_close_ep(ep);
+        return;
+    }
+
+    if (!ucs_socket_is_connected(fd)) {
+        ucs_debug("fd %d is not connected", fd);
+        /* coverity[check_return] */
+        ucs_async_modify_handler(fd, 0);
+        return;
+    }
+
+    switch (ep->state) {
```

Comment:
minor: switch for one state looks weird, i'd use `if` instead

Reply:
there are many more cases to come in the next prs. but ok, i'll change to 'if' for now.

Reply:
no problem then, can leave it as it is